File: src/stores/item.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import type { Item } from "@/types/item";
import { useTauriCommand } from "@/composables/useTauriCommand";
import { useToastStore } from "@/stores/toast";

export const useItemStore = defineStore("item", () => {
  const { call } = useTauriCommand();
  const items = ref<Item[]>([]);
  const loading = ref(false);

  async function fetchItemsByCategory(categoryId: number) {
    loading.value = true;
    try {
      items.value = await call<Item[]>("get_items_by_category", { categoryId });
    } finally {
      loading.value = false;
    }
  }

  async function createItem(
    categoryId: number,
    name: string,
    itemType: string,
    path: string,
    itemArgs?: string,
    workingDir?: string
  ) {
    await call("create_item", { categoryId, name, itemType, path, arguments: itemArgs || null, workingDir: workingDir || null });
  }

  async function updateItem(
    id: number,
    updates: {
      name?: string;
      itemType?: string;
      path?: string;
      arguments?: string;
      workingDir?: string;
    }
  ) {
    await call("update_item", {
      id,
      name: updates.name,
      itemType: updates.itemType,
      path: updates.path,
      arguments: updates.arguments,
      workingDir: updates.workingDir,
    });
  }

  async function moveItem(id: number, targetCategoryId: number) {
    await call("move_item", { id, targetCategoryId });
  }

  async function launchItem(id: number) {
    try {
      await call("launch_item", { id });
    } catch (e: unknown) {
      const toast = useToastStore();
      const msg =
        e && typeof e === "object" && "message" in e
          ? (e as { message: string }).message
          : "启动失败，请检查路径是否正确";
      toast.error(msg);
    }
  }

  async function deleteItem(id: number) {
    await call("delete_item", { id });
  }

  async function batchDeleteItems(ids: number[]) {
    await call("batch_delete_items", { ids });
  }

  async function togglePinItem(id: number) {
    await call("toggle_pin_item", { id });
  }

  async function reorderItems(orders: { id: number; sort_order: number }[]) {
    await call("reorder_items", { orders });
  }

  return {
    items,
    loading,
    fetchItemsByCategory,
    createItem,
    updateItem,
    moveItem,
    launchItem,
    deleteItem,
    batchDeleteItems,
    togglePinItem,
    reorderItems,
  };
});

File: src-tauri/src/commands/item.rs
use crate::db::connection::DbState;
use crate::db::models::{AppError, Item};
use tauri::State;

#[tauri::command]
pub fn get_items_by_category(
    db: State<'_, DbState>,
    category_id: i64,
) -> Result<Vec<Item>, AppError> {
    crate::services::item_service::get_items_by_category(&db, category_id)
}

#[tauri::command]
pub fn create_item(
    db: State<'_, DbState>,
    category_id: i64,
    name: String,
    item_type: String,
    path: String,
) -> Result<Item, AppError> {
    crate::services::item_service::create_item(&db, category_id, &name, &item_type, &path)
}

#[tauri::command]
pub fn launch_item(db: State<'_, DbState>, id: i64) -> Result<(), AppError> {
    crate::services::item_service::launch_item(&db, id)
}

#[tauri::command]
pub fn move_item(
    db: State<'_, DbState>,
    id: i64,
    target_category_id: i64,
) -> Result<(), AppError> {
    crate::services::item_service::move_item(&db, id, target_category_id)
}

#[tauri::command]
pub fn reorder_items(
    db: State<'_, DbState>,
    orders: Vec<(i64, i32)>,
) -> Result<(), AppError> {
    crate::services::item_service::reorder_items(&db, &orders)
}

#[tauri::command]
pub fn delete_item(db: State<'_, DbState>, id: i64) -> Result<(), AppError> {
    crate::services::item_service::delete_item(&db, id)
}

#[tauri::command]
pub fn toggle_pin_item(db: State<'_, DbState>, id: i64) -> Result<(), AppError> {
    crate::services::item_service::toggle_pin_item(&db, id)
}
